File: ios/renderer/ListItemRenderer.m
#import "ListItemRenderer.h"
#import "ASTNodeWrapper.h"
#import "RenderContext.h"
#import "StyleAttributes.h"
#import "StyleConfig.h"

@implementation ListItemRenderer

- (void)renderNode:(ASTNodeWrapper *)node
              into:(NSMutableAttributedString *)output
           context:(RenderContext *)context {
  NSMutableDictionary *attrs = [context.currentAttributes mutableCopy];
  [StyleAttributes applyParagraphPropertiesFromStyle:context.styleConfig.base
                                             toAttrs:attrs];
  [StyleAttributes applyStyle:context.styleConfig.listItem toAttrs:attrs];

  // md4c omits MD_BLOCK_P for items in tight lists, so a list item's
  // text can be rendered inline without a trailing newline — which
  // would make the next item's bullet land on the same line. Force a
  // preceding newline whenever we're not at the very start of the
  // output buffer.
  if (output.length > 0) {
    unichar lastChar = [output.string characterAtIndex:output.length - 1];
    if (lastChar != '\n') {
      [output appendAttributedString:
          [[NSAttributedString alloc] initWithString:@"\n"
                                          attributes:attrs]];
    }
  }

  // Build indent
  NSString *indent = @"";
  for (NSInteger i = 1; i < context.listDepth; i++) {
    indent = [indent stringByAppendingString:@"    "];
  }

  // Build bullet/number
  NSString *bullet;
  if (node.isTaskItem) {
    bullet = node.taskChecked ? @"[x] " : @"[ ] ";
  } else if (context.currentListIsOrdered) {
    // Left-pad the number with figure spaces (U+2007 — a digit-width
    // whitespace character in most fonts) so the periods line up for
    // mixed-width markers like 1./10./11.
    NSInteger number = context.orderedListIndex;
    NSInteger digits = 1;
    NSInteger tmp = MAX(1, number);
    while (tmp >= 10) {
      digits++;
      tmp /= 10;
    }
    NSInteger padCount =
        MAX(0, context.currentListMaxMarkerDigits - digits);
    NSMutableString *padding = [NSMutableString new];
    for (NSInteger i = 0; i < padCount; i++) {
      [padding appendString:@"\u2007"];
    }
    bullet = [NSString stringWithFormat:@"%@%ld. ", padding, (long)number];
    context.orderedListIndex++;
  } else {
    NSArray *bullets = @[ @"\u2022 ", @"\u25E6 ", @"\u25AA " ];
    NSInteger bulletIndex = (context.listDepth - 1) % bullets.count;
    bullet = bullets[bulletIndex];
  }

  NSString *prefix = [indent stringByAppendingString:bullet];

  // Bullet gets its own style on top of the list item attrs
  NSMutableDictionary *bulletAttrs = [attrs mutableCopy];
  [StyleAttributes applyStyle:context.styleConfig.listBullet toAttrs:bulletAttrs];

  [output appendAttributedString:
      [[NSAttributedString alloc] initWithString:prefix attributes:bulletAttrs]];

  [context pushAttributes:attrs];
  [context renderChildren:node into:output];
  [context popAttributes];

  // Ensure each list item ends with a newline
  if (output.length > 0) {
    unichar lastChar = [output.string characterAtIndex:output.length - 1];
    if (lastChar != '\n') {
      [output appendAttributedString:
          [[NSAttributedString alloc] initWithString:@"\n"
                                          attributes:attrs]];
    }
  }
}

@end

File: android/src/main/java/com/alizahid/markdown/renderer/ListItemRenderer.kt
package com.alizahid.markdown.renderer

import android.text.SpannableStringBuilder
import com.alizahid.markdown.parser.AstNode
import com.alizahid.markdown.renderer.RenderContext.Companion.applyAttributes
import com.alizahid.markdown.renderer.RenderContext.Companion.mergeStyleAttrs

/**
 * Mirrors ios/renderer/ListItemRenderer.m. Emits an optional separating
 * newline, an indent + bullet/number prefix styled with listBullet over
 * the listItem cascade, then the item's children with listItem attrs
 * pushed, then a trailing newline.
 */
object ListItemRenderer : NodeRenderer {

  private val bulletGlyphs = arrayOf("• ", "◦ ", "▪ ")

  /** Digit-width space (U+2007) — pads ordered markers so periods align. */
  private const val FIGURE_SPACE = '\u2007'

  override fun render(node: AstNode, into: SpannableStringBuilder, ctx: RenderContext) {
    val itemStyle = ctx.styleConfig.listItem
    val itemAttrs = mergeStyleAttrs(itemStyle, ctx.currentAttributes())
    val bulletAttrs = mergeStyleAttrs(ctx.styleConfig.listBullet, itemAttrs)

    // md4c omits MD_BLOCK_P for items in tight lists, so an item's text
    // can end without a newline — force a separator so the next item's
    // bullet doesn't land on the same line.
    if (into.isNotEmpty() && into[into.length - 1] != '\n') into.append('\n')

    val indent = "    ".repeat((ctx.listDepth - 1).coerceAtLeast(0))

    val bullet: String = if (ctx.currentListIsOrdered) {
      val number = ctx.orderedListIndex
      val digits = digitsOf(number)
      val padCount = (ctx.currentListMaxMarkerDigits - digits).coerceAtLeast(0)
      ctx.orderedListIndex = number + 1
      buildString {
        repeat(padCount) { append(FIGURE_SPACE) }
        append(number)
        append(". ")
      }
    } else {
      bulletGlyphs[((ctx.listDepth - 1).coerceAtLeast(0)) % bulletGlyphs.size]
    }

    val itemStart = into.length
    into.append(indent)
    into.append(bullet)
    applyAttributes(bulletAttrs, into, itemStart, into.length)

    ctx.pushAttributes(itemAttrs)
    ctx.renderChildren(node, into)
    ctx.popAttributes()

    // lineHeight / textAlign for the item, cascading base like iOS
    // (applyParagraphPropertiesFromStyle:base + applyStyle:listItem).
    applyBlockParagraphProps(itemStyle, ctx.styleConfig.base, into, itemStart, into.length)

    if (into.isNotEmpty() && into[into.length - 1] != '\n') into.append('\n')
  }

  private fun digitsOf(n: Int): Int {
    var v = maxOf(1, n)
    var d = 1
    while (v >= 10) { d++; v /= 10 }
    return d
  }
}
